tools/cross_runner/src/mha.h

```diff
@@ -0,0 +1,490 @@
+                    tensor_out_desc_.Sizes = output_dims.data();
+
+                    tensor_out_desc_.Strides = nullptr; 
+                    tensor_out_desc_.TotalTensorSizeInBytes = DMLCalcBufferTensorSize(
+                        tensor_out_desc_.DataType,
+                        tensor_out_desc_.DimensionCount,
+                        tensor_out_desc_.Sizes,
+                        tensor_out_desc_.Strides);
+                    tensor_out_desc_.GuaranteedBaseOffsetAlignment = 0; 
+                }
+
+                DML_TENSOR_DESC output_desc{};
+                output_desc.Desc = &tensor_out_desc_;
+                output_desc.Type = DML_TENSOR_TYPE_BUFFER;
+
+                DML_TENSOR_DESC stacked_qkv_tensor_desc = {};
+                stacked_qkv_tensor_desc.Desc = &tensor_stacked_qkv_desc_;
+                stacked_qkv_tensor_desc.Type = DML_TENSOR_TYPE_BUFFER;
+                
+                DML_MULTIHEAD_ATTENTION_OPERATOR_DESC desc = {};
+                desc.StackedQueryKeyValueTensor = &stacked_qkv_tensor_desc;
+                desc.OutputTensor = &output_desc;
+                desc.Scale = 0.001;
+                desc.MaskFilterValue = -10000000;
+                desc.HeadCount = 8;
+                desc.MaskType = DML_MULTIHEAD_ATTENTION_MASK_TYPE_NONE;
+
+                DML_OPERATOR_DESC dml_operator_desc{};
+                dml_operator_desc.Type = DML_OPERATOR_MULTIHEAD_ATTENTION;
+                dml_operator_desc.Desc = &desc;
+
+                throw_if_failed(dml_device->CreateOperator(
+                    &dml_operator_desc, IID_PPV_ARGS(dml_operator_.ReleaseAndGetAddressOf())), "create Multihead Attention operator");
+
+                DML_EXECUTION_FLAGS exec_flags = DML_EXECUTION_FLAG_NONE; //| DML_EXECUTION_FLAG_DESCRIPTORS_VOLATILE;
+                if (data_type == DML_TENSOR_DATA_TYPE_FLOAT16)
+                {
+                    exec_flags |= DML_EXECUTION_FLAG_ALLOW_HALF_PRECISION_COMPUTATION;
+                }
+
+                if (disable_mc)
+                {
+                    exec_flags |= DML_EXECUTION_FLAG_DISABLE_META_COMMANDS;
+                }
+
+                throw_if_failed(dml_device->CompileOperator(
+                    dml_operator_.Get(),
+                    exec_flags,
+                    IID_PPV_ARGS(dml_op_executor_.ReleaseAndGetAddressOf())), "create Multihead Attention compiled operator");
+
+                create_operator_impl();
+            }else{
+                assert(false && "Unsupported MHA type!");
+            }
+        }
+
+        dml::TensorDesc get_tensor_input_desc() const
+        {
+            return tensor_stacked_qkv_desc_;
+        }
+
+        dml::TensorDesc get_tensor_out_desc() const
+        {
+            return tensor_out_desc_;
+        }
+
+        void record_execute(IDMLCommandRecorder* dml_cmd_recorder, ID3D12GraphicsCommandList* cmd_list,
+            ID3D12Resource* resource_out, ID3D12Resource* resource_a)
```

Comment:
lets have more explicit resource naming, i.e. resource_qkv

Reply:
Do we need to hardcode this? Will it have difference meaning if we implement different mha type in the future?